app/src/main/java/org/mozilla/focus/home/HomeFragment.java

```diff
@@ -862,15 +862,21 @@ private UpdateHistoryWrapper(List<String> urls, WeakReference<Handler> handlerWe
 
         @Override
         public void accept(List<String> fileUris) {
+            QueryHandler.AsyncUpdateListener listener = result -> {
+                Handler handler = handlerWeakReference.get();
+                if (handler == null) {
+                    return;
+                }
+                Message message = handler.obtainMessage(MSG_ID_REFRESH);
+                handler.dispatchMessage(message);
+            };
             for (int i = 0 ; i < fileUris.size() ; i++) {
-                BrowsingHistoryManager.updateHistory(null, urls.get(i), fileUris.get(i));
-            }
-            Handler handler = handlerWeakReference.get();
-            if (handler == null) {
-                return;
+                if (i == fileUris.size() - 1) {
```

Comment:
Access to SQLite is [synchronous](https://www.androiddesignpatterns.com/2012/10/sqlite-contentprovider-thread-safety.html), so it is guaranteed that when the last entry is complete, all previous is also completed.